test/java/org/chromium/net/impl/CancelProofEnvoyStreamTest.java

```diff
@@ -0,0 +1,379 @@
+    mSendDataBlock.open();
+    thread.join(); // Wait for the Thread to die.
+
+    assertThat(mSendDataInvocationCount.get()).isOne();
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  @Test
+  public void readData() {
+    cancelProofEnvoyStream.setStream(mMockedStream);
+
+    cancelProofEnvoyStream.readData(1);
+
+    assertThat(mReadDataInvocationCount.get()).isOne();
+    assertThat(mCancelInvocationCount.get()).isZero();
+  }
+
+  @Test
+  public void readData_withoutStreamSet() {
+    assertThatThrownBy(() -> cancelProofEnvoyStream.readData(1))
+        .isInstanceOf(NullPointerException.class);
+  }
+
+  @Test
+  public void readData_afterCancel() {
+    cancelProofEnvoyStream.setStream(mMockedStream);
+    cancelProofEnvoyStream.cancel();
+
+    cancelProofEnvoyStream.readData(1);
+
+    assertThat(mReadDataInvocationCount.get()).isZero();
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  @Test
+  public void readData_postponedCancelGetsExecutedToo() throws Exception {
+    cancelProofEnvoyStream.setStream(mMockedStream);
+    mReadDataBlock.close(); // Following Thread will block on executing readData
+    Thread thread = new Thread(() -> cancelProofEnvoyStream.readData(1));
+    thread.start();
+    mStartLatch = new CountDownLatch(1); // Only one method to wait for.
+    mStartLatch.await();                 // Wait for the above Thread to enter the readData method.
+    cancelProofEnvoyStream.cancel();
+    mReadDataBlock.open();
+    thread.join(); // Wait for the Thread to die.
+
+    assertThat(mReadDataInvocationCount.get()).isOne();
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  @Test
+  public void cancel() {
+    cancelProofEnvoyStream.setStream(mMockedStream);
+
+    cancelProofEnvoyStream.cancel();
+
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  @Test
+  public void cancel_twice_executedOnlyOnce() {
+    cancelProofEnvoyStream.setStream(mMockedStream);
+
+    cancelProofEnvoyStream.cancel();
+    cancelProofEnvoyStream.cancel();
+
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  @Test
+  public void cancel_manyConcurrentThreads_executedOnlyOnce() throws Exception {
+    cancelProofEnvoyStream.setStream(mMockedStream);
+    Thread[] threads = new Thread[100];
+    for (int i = 0; i < threads.length; i++) {
+      threads[i] = new Thread(cancelProofEnvoyStream::cancel);
+    }
+    for (Thread thread : threads) {
+      thread.start();
+    }
+    // Wait for all the Threads to die.
+    for (Thread thread : threads) {
+      thread.join();
+    }
+
+    assertThat(mCancelInvocationCount.get()).isOne();
+  }
+
+  @Test
+  public void cancel_withoutStreamSet() {
+    // Does nothing because the "Concurrently Running Stream Operations Count" starts with "1".
+    // This is the desired outcome - the cancel is postponed. Once the stream is set, the next
+    // Stream Operation will invoke "cancel".
+    cancelProofEnvoyStream.cancel();
```

Comment:
Should this check that the cancel invocation count is 0?

Reply:
It is the `MockedStream` that can increment the cancel invocation count, and this test does not invoke `setStream`: it is pointless to verify it.

Added a comment.